src/Lean/Server/Watchdog.lean

```diff
@@ -240,10 +242,15 @@ section ServerM
   def startFileWorker (m : DocumentMeta) : ServerM Unit := do
     publishProgressAtPos m 0 (← read).hOut
     let st ← read
+    let mut args := st.args.toArray
+    if let some path := fileUriToPath? m.uri then
+      if let some (_, pkg) ← searchModuleNameAndPkgOfFileName path st.pkgSearchPath then
+        if let some pkgArgs := st.pkgArgs.find? pkg then
+          args := pkgArgs
```

Comment:
Note that in the case that came up on zulip (browsing std from mathlib), the "original" arguments are mathlib's and the package arguments are std's (std has no package options), so extending would cause the same issue as already exists.

Reply:
@digama0 As I mentioned in the Zulip thread, `moreServerArgs` is global. `moreLeanArgs` is the package-specific one. Thus, it should be a combination of the two.

Reply:
`moreLeanArgs` is also documented as only applying to the CLI build though, not the server (and the names back this interpretation up). Nothing about those names suggest that one is global and the other is package-specific, although one of them is on the package and the other is on the library.

Reply:
I implemented `moreGlobalServerArgs` and added `moreServerArgs` as indicated. Unfortunately, the implementation is not perfectly accurate because the server does not actually know the precise import structure of each package: if it sees something in the lib's folder then it assumes it is from the lib, even if it is not a transitive import of the root library. This is potentially a problem for packages like mathlib, because there are several libs (e.g. `Mathlib`, `Counterexamples`) that share the same root folder (the repo root), because the `Mathlib/` folder doesn't count as part of the `srcDir` and hence there is a lot of overlap between the directory roots for each lib.

As a result per-lib `moreServerArgs` doesn't really work in mathlib and other projects that have multiple libs with the same root directory. This isn't likely to be an issue for mathlib though since it can just set the `moreServerArgs` at the package level - it isn't (currently) attempting to set options differently per-lib. If it did want to do so, it would have to put them in separate folders, or we would have to revisit this logic to give the server more project structure information.

Reply:
@digama0 Thank you for the changes! Could you call `get-pkg`-args` (or, I guess `get-mod-args`) for the file each time before starting a worker to get the arguments of the that module? The would hopefully make per-lib configurations work correctly in all cases.  While it may introduce some overhead, restarting a file already takes some time (due to building imports) so I do not think additional overhead would noticeable..